mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala

```diff
@@ -157,6 +157,46 @@ class RankingMetrics[T: ClassTag](predictionAndLabels: RDD[(Array[T], Array[T])]
     }.mean()
   }
 
+  /**
+    * Compute the average recall of all the queries, truncated at ranking position k.
+    *
+    * If for a query, the ranking algorithm returns n results, the recall value will be
+    * computed as #(relevant items retrieved) / #(ground truth set). This formula
+    * also applies when the size of the ground truth set is less than k.
+    *
+    * If a query has an empty ground truth set, zero will be used as recall together with
+    * a log warning.
+    *
+    * See the following paper for detail:
+    *
+    * IR evaluation methods for retrieving highly relevant documents. K. Jarvelin and J. Kekalainen
+    *
+    * @param k the position to compute the truncated recall, must be positive
+    * @return the average recall at the first k ranking positions
+    */
+  @Since("2.5.0")
+  def recallAt(k: Int): Double = {
+    require(k > 0, "ranking position k should be positive")
+    predictionAndLabels.map { case (pred, lab) =>
```

Comment:
This is so similar to precision at k that they should be refactored to use one common private implementation method for 90% of it.

Reply:
Thanks, I refactored the code.